### ISSUE TITLE
Could we override the module/theme handler services like Hook manager does?

### ISSUE THREAD

**Issue by Ambient-Impact** [pdenooijer/hook_event_dispatcher#112]
When I was first looking into object-oriented hooks for Drupal 8, I found [Hook manager](https://www.drupal.org/project/hook_manager). It seems more of a proof of concept, and uses plug-ins rather than events, but it does what I feel is a really clever thing to avoid having to implement each procedural hook by [extending the ```module_handler``` service](https://git.drupalcode.org/project/hook_manager/blob/8.x-1.x/src/HookManagerServiceProvider.php) with [its own class](https://git.drupalcode.org/project/hook_manager/blob/8.x-1.x/src/Service/HookManagerModuleHandler.php) that first executes ```parent::invokeAll()``` to run procedural hooks and then has its own manager executing any plug-in classes that implement matching hooks.

@pdenooijer, could we do something similar by extending the module and theme handler services to trigger events directly rather than implementing every hook ourselves? Is there something that I've overlooked that would make implementing this difficult for ```hook_event_dispatcher```?

**Comment by pdenooijer** [comment]
The problem with this is that you will have very generic events that can not have specific methods and have to have a way to get the given arguments out of the hook somehow which is not really easy to use.

I have thought about this already a bit, but this generic event thing is my biggest concern at the moment.

Al though we could use something similar in this module by [decorating](https://symfony.com/doc/current/service_container/service_decoration.html) the module_handler, so we can set the priority of our events a bit better having the ability to remove all the hooks in this project.

**Comment by Ambient-Impact** [comment]
Hmm, I see. Yeah, we'd definitely want to avoid being stuck with generic events. Decorating the ```module_handler``` service seems like it could be the way to go, although I'm still trying to wrap my brain around how to keep the flexibility of the events we already have but make this work for almost any conceivable hook. Drupal's plug-in system is great at this, but that doesn't help with events.

One possibility that occurs to me is creating some sort of registry, like the theme registry or how [field formatters work](https://api.drupal.org/api/drupal/core%21modules%21field%21field.api.php/group/field_formatter/8.8.x): each event can have a registered event class for a given hook, which should theoretically allow the same level of flexibility with the current events while allowing us to remove the procedural functions. I'm not sure how we'd register stuff in said registry, though. Could we use annotations? Do those only work with Drupal's plug-in system? What do you think?

**Comment by Ambient-Impact** [comment]
@pdenooijer What do you think about the registry approach? Modules that provide hook events could have a ```<module name>.hooks.yml``` or something like that which would list the default event classes for hooks, in a similar way as services or route controllers.

**Comment by pdenooijer** [comment]
The hooks.yml approach requires us to discover if this file exists in all modules, I don't really like that approach. Based on that idea I was thinking more in the way of creating a tagged service in each module that would add the events to the handler.

**Comment by Ambient-Impact** [comment]
While I can understand that a tagged service would make this simpler for us, would it make the developer experience better? Core already has the established pattern of ```<module name>.<type>.yml```, and this declarative approach feels a lot cleaner and simpler. That said, we could also provide a tagged service in addition to that, if it serves some advanced use cases.

I do agree that having to discover and manage YML files ourselves seems like more work than we should be doing, but after a bit of searching through core, I've found that there's a [```YamlDiscovery``` class](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Discovery%21YamlDiscovery.php/class/YamlDiscovery/8.8.x) that seems to handle this exact problem for us. It's used in [```RouteBuilder::getRouteDefinitions()```](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Routing%21RouteBuilder.php/function/RouteBuilder%3A%3AgetRouteDefinitions/8.8.x) like so:

```
  $discovery = new YamlDiscovery('routing', $this->moduleHandler
    ->getModuleDirectories());
  return $discovery
    ->findAll();
```

Does that change how you feel about it?

**Comment by pdenooijer** [comment]
> While I can understand that a tagged service would make this simpler for us, would it make the developer experience better?

Why not? The developer would subscribe to the event right? If the event will construct the string where to listen to you can just use that.

**Comment by Ambient-Impact** [comment]
I wonder if we're talking about slightly different things? Can you give me an example of how you see a developer defining a hook?

**Comment by pdenooijer** [comment]
What I mean is that with a tagged service there is more freedom to create hooks that are defined depending on the entities that are in the system. For instance a newsletter entity would have an `newsletter_update` hook. I want to see this behaviour possible as I find that one of the current drawbacks of this module. Dynamic hooks should be possible in the events as well. I don't think that will be possible with YAML files or do you have an idea how to get that functionality?

**Comment by Ambient-Impact** [comment]
Are you referring to the original invocation of a hook (i.e. the module handler doing so), or the implementation of/subscribing to the hook? I'm only talking about the former; I'm not proposing any changes to how a developer should subscribe to a hook at all.

Let me try to illustrate with a concrete example. Right now, we have to have the procedural function ```core_event_dispatcher_form_alter()``` (implementation of ```hook_form_alter()```) to dispatch ```FormAlterEvent```:

```
function core_event_dispatcher_form_alter(array &$form, FormStateInterface $formState, string $formId) {
  ...
}
```

If we were to extend the module handler so that we can intercept the invocation of ```hook_form_alter``` and remove the ```core_event_dispatcher_form_alter()``` function, we have a problem because now our module handler doesn't know what event to map to that hook when Drupal invokes it, correct? What I propose is that we create a YAML file that maps the hook name to the event object, so that our module handler knows which event object to use:

```core_event_dispatcher.hooks.yml```:
```
hooks:
  form_alter:
    class: \Drupal\core_event_dispatcher\Event\Form\FormAlterEvent
```

As for how we'd handle dynamic hooks - such as for ```hook_form_FORM_ID_alter()``` - we'll figure something out in #111, but I see the same basic concept working for that as well:

```core_event_dispatcher.hooks.yml```:
```
hooks:
  form_id_alter:
    class: \Drupal\core_event_dispatcher\Event\Form\FormIdAlterEvent
```

Event constructors already accept the parameters from their procedural hook counterparts that dispatch them, so we may not need to change much there.

Let me know if that makes things a little clearer to you.

**Comment by pdenooijer** [comment]
I am talking about the former as well, but then including the ability to have dynamic hooks in there. I really would like to have that functionality.

**Comment by Ambient-Impact** [comment]
Just to make sure I'm understanding correctly, when you refer to dynamic hooks, do you hooks like ```hook_form_FORM_ID_alter()```, etc.?

**Comment by pdenooijer** [comment]
Yes, that's right 🙂. Mainly what that to improve the preprocess further and other dynamic hooks a that are used more often.

I did some work in the `feature/decorate_module_handler` branch figuring out how to hook into the ModuleHandler it self. To bad that class is not SOLID at all and it has two ways to invoke a module hook.

I am still trying to figure out how to setup a way to send out the events, one of the options is:

Trigger an event with weight 0 to run the default hooks, that way there is a possibility to either have an event before that and after that with a different weight. I do want to keep this slim and agile. For unknown hooks fall back to the hook system.

**Comment by Ambient-Impact** [comment]
Yeah, it seems like we could probably just merge the preprocess and dynamic hook functionality to reduce code duplication. I took a look yesterday at the WIP class, though I don't have much of an opinion at the moment.

> I am still trying to figure out how to setup a way to send out the events, one of the options is:

> Trigger an event with weight 0 to run the default hooks, that way there is a possibility to either have an event before that and after that with a different weight. I do want to keep this slim and agile. For unknown hooks fall back to the hook system.

That makes sense, yeah. Having it delegate that to the event system seems like the way to go and is also super meta. Would that be instead of or potentially working with my proposed way of declaratively defining new hooks? We could create a service whose single purpose is to gather hook definitions and pass that as a dependency to the new module handler.

**Comment by pdenooijer** [comment]
> That makes sense, yeah. Having it delegate that to the event system seems like the way to go and is also super meta. Would that be instead of or potentially working with my proposed way of declaratively defining new hooks? We could create a service whose single purpose is to gather hook definitions and pass that as a dependency to the new module handler.

For sure multiple classes that do a very specific job and the ModuleHandler will only be the overwrite and fallback for the hook system 🙂.

**Comment by Ambient-Impact** [comment]
🚀🎉👍
